lte/gateway/c/session_manager/SessionStateEnforcer.cpp

```diff
@@ -136,15 +134,252 @@ void SessionStateEnforcer::handle_session_init_rule_updates(
+{
+   auto map_element = session_map.find(imsi);
+   if(map_element == session_map.end()) {
+      MLOG(MERROR) << "No session found in SessionMap for IMSI " << imsi
+                   << " during terminating session_id " << session_id
+		   << "skipping termination";
+      return;
+   }
+
+   SessionStateUpdateCriteria& uc = session_update[imsi][session_id];
+
+   for (auto session_element = map_element->second.begin(); 
+        session_element != map_element->second.end();
+        ++session_element) {
+      if ((*session_element)->get_session_id() == session_id) {
+         uc.is_session_ended = true;
+	 /*Removing session from map*/
+         map_element->second.erase(session_element--);
+         MLOG(MINFO) << "Successfully terminated session " << session_id;
+         /* If it is last session terminated and no session left for this IMSI
+	  * remove the imsi as well
+	  */
+         if (map_element->second.size() == 0) {
+            session_map.erase(imsi);
+            MLOG(MINFO) << "All sessions terminated for IMSI " << imsi;
+         }
+         return;
+      }
+   }//end of for loop
+   //reached end of map and no session found.
+   MLOG(MINFO) << "Could not find session " << session_id
+               << "to complete termination of IMSI " << imsi;
+   return;
+}
+
+/*Removing all associated rules to session*/
+void SessionStateEnforcer::m5g_remove_rules_from_session_for_termination(
+      const std::string& imsi, 
+      const std::unique_ptr<SessionState>& session,
+      const std::string& dnn, SessionStateUpdateCriteria& uc)
+{
+  // remove all rules;
+  MLOG(MINFO) << "Will be removing all associated rules of IMSI " << imsi;
+  session->remove_all_rules();
+  auto ip_addr = session->get_config()
+                     .rat_specific_context.m5gsm_session_context()
+                     .pdu_address()
+                     .redirect_server_address();
+  //Update to UPF
+  SessionState::SessionInfo sess_info;
+  sess_info.imsi       = imsi;
+  sess_info.ip_addr    = ip_addr;
+  session->sess_infocopy(&sess_info);
+  pipelined_client_->set_upf_session(sess_info, call_back_void_upf);
+   return;
+}
+
+/* This function will change the state of respective PDU session,
+ * upon receving message or notification from UPF
+ * Defined to handle ACTIVE or INACTIVE state of session
+ * based on UPF message to SMF
+ * imsi - from UPF handler to find respective SessionMap
+ * seid - session context id to find respective session
+ * new_state - state change required w.r.t. UPF message
+ */
+void SessionStateEnforcer::m5g_update_session_state_message_from_upf(
+		const std::string& imsi, const std::string& session_id,
+		SessionFsmState new_state)
+{
+  auto session_map    = session_store_.read_sessions({imsi});
+  auto session_update = 
+	     SessionStore::get_default_session_update(session_map);
+  auto map_element = session_map.find(imsi);
+  if(!(map_element == session_map.end())) {

```

Comment:
If session is not found we don't log anything. Maybe it is better to do 
```
if(map_element == session_map.end()){
	MLOG(MWARNING) << sessionid << " session not found while m5g_update_session_state_message_from_upf " 
	return;
}

```

Reply:
Adapted with little change
 MLOG(MWARNING) << "Session id (" << session_id << ") not found to update"
                   << " SessionStore based on UPF message";